pkg/receive/handler.go

```diff
@@ -1310,17 +1357,21 @@ func (p *peerGroup) getConnection(ctx context.Context, addr string) (WriteableSt
 	defer p.m.Unlock()
 	// Make sure that another caller hasn't created the connection since obtaining the write lock.
 	c, ok = p.connections[addr]
-	if ok {
+	if ok && c.isReady() {
 		return c, nil
 	}
+	level.Debug(p.logger).Log("msg", "dialing peer", "addr", addr)
 	conn, err := p.dialer(ctx, addr, p.dialOpts...)
 	if err != nil {
 		p.markPeerUnavailableUnlocked(addr)
 		dialError := errors.Wrap(err, "failed to dial peer")
 		return nil, errors.Wrap(dialError, errUnavailable.Error())
 	}
-
-	p.connections[addr] = newPeerWorker(conn, p.forwardDelay, p.asyncForwardWorkersCount)
+	if !ok {
+		p.connections[addr] = newPeerWorker(conn, p.forwardDelay, p.asyncForwardWorkersCount)
+	} else {
+		p.connections[addr].cc = conn
```

Comment:
What if, in reality (in contrast to the unit test), re-dialing still sees the stale IP because of some weird interaction between gRPC's DNS resolver and Kubernetes networking (CoreDNS)?

Reply:
plan to rever this one.